src/raft.rs

```diff
@@ -1222,102 +1256,20 @@ impl<T: Storage> Raft<T> {
-                "!ConfChange::has_start_index()".into(),
-            ));
-        };
-
-        self.set_pending_membership_change(conf_change.clone());
-        let max_inflights = self.max_inflight;
-        self.mut_prs()
-            .begin_membership_change(configuration, Progress::new(1, max_inflights))?;
+    fn begin_membership_change(&mut self, cs: &ConfState) -> Result<()> {
+        debug!("{} begin membership change to {:?}", self.tag, cs);
+        let progress = Progress::new(1, self.max_inflight);
+        self.mut_prs().begin_membership_change(cs, progress)?;
         Ok(())
     }
 
-    /// Apply a `FinalizeMembershipChange` variant `ConfChange`.
-    ///
-    /// > **Note:** This is an experimental feature.
-    ///
-    /// When a Raft node applies this variant of a configuration change it will finalize the
-    /// transition begun by [`begin_membership_change`].
-    ///
-    /// Once this is called the Raft will no longer have two, possibly overlapping, cooperating
-    /// qourums.
-    ///
-    /// # Errors
-    ///
-    /// * This Raft is not in a configuration change via `begin_membership_change`.
-    /// * `ConfChange.change_type` is not a `FinalizeMembershipChange`.
-    /// * `ConfChange.configuration` value should not exist.
-    /// * `ConfChange.start_index` value should not exist.
     #[inline(always)]
-    pub fn finalize_membership_change(&mut self, conf_change: &ConfChange) -> Result<()> {
-        if conf_change.get_change_type() != ConfChangeType::FinalizeMembershipChange {
-            return Err(Error::ViolatesContract(format!(
-                "{:?} != BeginMembershipChange",
-                conf_change.get_change_type()
-            )));
-        }
-        if conf_change.has_configuration() {
-            return Err(Error::ViolatesContract(
-                "ConfChange::has_configuration()".into(),
-            ));
-        };
-        let leader_in_new_set = self
-            .prs()
-            .next_configuration()
-            .as_ref()
-            .map(|config| config.contains(self.leader_id))
-            .ok_or_else(|| Error::NoPendingMembershipChange)?;
-
-        // Joint Consensus, in the Raft paper, states the leader should step down and become a
-        // follower if it is removed during a transition.
-        if !leader_in_new_set {
-            let last_term = self.raft_log.last_term();
-            if self.state == StateRole::Leader {
-                self.become_follower(last_term, INVALID_ID);
-            } else {
-                // It's no longer safe to lookup the ID in the ProgressSet, remove it.
-                self.leader_id = INVALID_ID;
-            }
-        }
-
+    fn finalize_membership_change(&mut self) -> Result<()> {
+        assert!(self.is_in_membership_change());
+        // Here we can't call `become_follower` because we need to bcast the entry later.
+        // Call that function will cause wrong `next_idx`s.
```

Comment:
```suggestion
        // Calling that function will cause wrong `next_idx`s.
```

Really great comment/explanation. 👍